docs(learncss): Add Backgrounds guide (#6205)
GoogleChrome/web.dev: src/site/content/en/learn/css/backgrounds/index.md

---
title: Backgrounds
description: >

>   In this module learn the ways you can style backgrounds of CSS boxes using CSS

Reviewer: ```suggestion
  In this module learn the ways you can style backgrounds of boxes using CSS
```

audio:
  title: The CSS Podcast - 053: Background
  src: https://traffic.libsyn.com/secure/thecsspodcast/TCP052_v1.mp3?dest-id=1891556
  thumbnail: image/foR0vJZKULb5AGJExlazy1xYDgI2/ECDb0qa4TB7yUsHwBic8.png
authors:
 - lozandier
date: 2021-9-15

---

## Backgrounds

Behind every CSS box is a specialized layer called the background layer. CSS provides a variety of ways to make meaningful changes to it–including allowing multiple backgrounds.

Background layers are furthest from the user, rendered behind the contents of a box starting from its `padding-box` region. This enables the background layer to not overlap with borders at all.

{% Codepen {
  user: 'argyleink',
  id: 'BaLedvd',
  tab: 'result'
} %}

> ## Background Color

Reviewer: ```suggestion
## Background color
```

One of the simplest effects you can apply to a background layer is setting the [color](/learn/css/color/). The initial value of `background-color` is `transparent`, which allows the contents of a parent to be visible. A valid color set on a background layer sits behind other things painted on that element.

{% Codepen {
  user: 'web-dot-dev',
  id: 'GRvqQZZ',
  tab: 'result'
} %}

> ## Background Image

Reviewer: ```suggestion
## Background images
```

On top of the `background-color` layer, you can add a background image, using the `background-image` property. A `background-image` accepts the following:

> * An image URL or [data URI](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs) using the `url` CSS function.

Reviewer: ```suggestion
* An image URL or [data URI](https://developer.mozilla.org/docs/Web/HTTP/Basics_of_HTTP/Data_URIs) using the `url` CSS function.
```

* An image dynamically created by a gradient CSS function.

> ### Setting a background-image with the `url` CSS function.

Reviewer: ```suggestion
### Setting a background-image with the `url` CSS function.

```

Reviewer: ```suggestion
### Setting a background-image with the `url` CSS function
```

{% Codepen {
  user: 'web-dot-dev',
  id: 'JjJNNro',
  tab: 'result'
} %}

### CSS gradient backgrounds

Several [gradient](/learn/css/gradients) CSS functions exist to allow you to generate a background-image, when passed two or more colors.

Regardless of which gradient function is used, the resulting image is [intrinsically sized](/learn/css/box-model/#content-and-sizing) to match the amount of space available.

Demo showing example of applying a background-image using Gradient functions:
{% Codepen {user: 'web-dot-dev', id: 'oNeLZWa', tab: 'css,result' } %}

> ## Background Repeat

Reviewer: ```suggestion
## Repeating background images
```

By default, background images repeat horizontally and vertically to fill the entire space of the background layer.

Change this by using the `background-repeat` property with one of the following values:

`repeat`: the image repeats within the space available, cropping as necessary.
`round`: the image repeats horizontally and vertically to fit as many instances into the space available, without cropping, compressing, or stretching it.

> `space`: the image repeats horizontally and vertically to fit as many instances within the space available without cropping—spacing out instances of the image as needed. Repeating images touch the edges of the space a background layer occupies, with white space evenly distributed between them:

Reviewer: ```suggestion
`space`: the image repeats horizontally and vertically to fit as many instances within the space available without cropping—spacing out instances of the image as needed. Repeating images touch the edges of the space a background layer occupies, with white space evenly distributed between them.
```

The `background-repeat` property allows you to set the behavior for the x and y axis independently. The first parameter sets the horizontal repeat behavior, and the second parameter sets the vertical repeat behavior.

If you use a single value, it will be applied to both the horizontal and vertical repeats.

The shorthand also has convenient one-word options to make your intent clearer.

> `repeat-x`  repeats an image only horizontally; this is equivalent to `repeat no-repeat`.

Reviewer: ```suggestion
`repeat-x` repeats an image only horizontally; this is equivalent to `repeat no-repeat`.
```

The following demo demonstrates these capabilities of the `background-repeat` property

{% Codepen {
  user: 'web-dot-dev',
  id: 'KKvMmjb',
  tab: 'result'
} %}

> ## Background Position

Reviewer: ```suggestion
## Background position
```

You may have noticed when some images on the Web are styled with a `background-repeat: no-repeat` declaration, such images are displayed top left of their container.

> The initial position of background images is top left. . The `background-position` allows you to change this behavior by offsetting the image position.

Reviewer: ```suggestion
The initial position of background images is top left. The `background-position`  property allows you to change this behavior by offsetting the image position.
```

As with `background-repeat`,  the `background-position` property allows you to position images along the x and y axis independently with two values by default.

When CSS lengths and percentages are used, the first parameter corresponds to the horizontal axis while the second parameter corresponds to the vertical axis.

When keywords are only used the order of the keywords does not matter:

> 
> 
> 

Reviewer: ```suggestion

```

{% Codepen {
  user: 'web-dot-dev',
  id: 'YzxWQqx',
  tab: 'result'
} %}

In addition to its default two parameter form and one parameter form; the `background-position` property also accepts up to four parameters;

When three or four parameters are used, a CSS length or percentage must be preceded by the `top`, `left`, `right`, or `bottom` keywords in order for the browser to calculate which edge of the CSS box the offset should originate from.

When three parameters are used, a CSS length or value can be the second or third parameter with the other two being keywords; the keyword it succeeds will be used to determine the edge the CSS length or value corresponds to being the offset of. The offset of the other keyword specified is set to 0.

{% Compare ‘better’ %}
```css
background-position: bottom 88% right;
```
{% endCompare %}

{% Compare ‘better’ %}
```css
background-position: right bottom 88%;
```
{% endCompare %}

{% Compare ‘better %}
```css
background-position: 88% bottom right;
```
{% CompareCaption %}

> CSS length value must be preceded by the 'top', 'right', 'bottom', or 'left keywords when using 3 or more parameters

Reviewer: ```suggestion
CSS length value must be preceded by the 'top', 'right', 'bottom', or 'left keywords when using 3 or more parameters.
```

```css
background-position: right 33% bottom 88%;
```
{% endCompare %}

{% Compare ‘better’ %}
```css
background-position: 88% 33% bottom left;
```
{% CompareCaption%}
 CSS length value must be preceded by the 'top', 'right', 'bottom', or 'left keywords when using 3 or more parameters
{% endCompareCaption%}
{% endCompare %}

If `background-position: top left 20%`is applied to a CSS background image, the image is placed at the top of the box, the `20%` value represents a 20% offset from the left of the box (on the x axis).

 If `background-position: top 20% left` is applied to a CSS background image, the 20% value represents a 20% offset from the top of the CSS box (on the y axis), and the image is placed at the left of the box.

When four parameters are used, the two keywords are paired with two values corresponding to an offset against the origins of each keyword specified. If `background-position: bottom 20% right 30%` is applied to a background-image, the background-image is positioned 20% from the bottom, and 30% from the right of the CSS box.

The following demo demonstrates this behavior:
{% Codepen {
  user: 'web-dot-dev',
  id: 'porbwrM',
  tab: 'result'
} %}

> Here are more examples of using the `background-position` property using a mix of CSS and keyword values:

Reviewer: There don't seem to be any linked examples?

> {% Callout ‘definition %}
> TODO: Decide to link to W3C spec or MDN To learn more about the nuances related to positioning backgrounds, check out MDN’s guide on the property](https://developer.mozilla.org/en-US/docs/Web/CSS/background-position)
> {% endCallout %}

Reviewer: ```suggestion
{% Aside %}
To learn more about the nuances related to positioning backgrounds, check out [`background-position` on MDN](https://developer.mozilla.org/docs/Web/CSS/background-position).
{% endAside %}
```

The `background-size` property accepts the following keywords:

* `auto`: When used independently, the background image is sized based on its intrinsic width and height; when `auto` is used alongside another CSS value for the width (first parameter) or height (second parameter), the dimension set to `auto` is sized as needed to maintain the natural aspect ratio of the image. This is the default behavior of the `background-size` property

* `cover`: Covers the entire area of the background layer. This may mean the image is stretched or cropped.

* `contain`:  Sizes the image to fill the space  without stretching or cropping. As a result, empty space can remain that will cause the image to repeat, unless `background-repeat` is set to `no-repeat`.

The latter 2 are intended to be used in a standalone fashion without another parameter.

The following demo demonstrates these keywords in action:

Demo demonstrating applying these keywords to `background-size`:
{% Codepen {
  user: 'web-dot-dev',
  id: 'YzxWQYY',
  tab: 'result'
} %}

> ## Background Attachment

Reviewer: ```suggestion
## Background attachment

```

The `background-attachment` property enables you to modify the fixed position behavior of background images (images part of a background layer) once the layer is visible on a screen.

It accepts 3 keywords: `scroll`, `fixed`, and `local`.

> * `scroll`: The default behavior of the `background-attachment` property. When more space is needed, the images move with that space within the background layer determined by the bounds of the CSS box.

Reviewer: ```suggestion
The default behavior of the `background-attachment` property is the initial value of `scroll`. When more space is needed, the images move with that space within the background layer determined by the bounds of the CSS box.
```

{% Codepen {
  user: 'web-dot-dev',
  id: 'mdwwzOe',
  tab: 'result'
} %}

> * `fixed`: fixes the position of background images to the viewport.

Reviewer: ```suggestion
Using the value `fixed` fixes the position of background images to the viewport.
```

Once the space of the background layer images originally takes up needs to be scrolled (or rendered) offscreen, images within the background layer stay fixed in the original position the background layer enabled them to be until the entire layer is scrolled off screen by the viewport.

{% Codepen {
  user: 'web-dot-dev',
  id: 'MWoozvN',
  tab: 'result'
} %}

> `local` : The use of this keyboard enables the position of background images to be fixed relative to the element’s contents. Background images now move along the space they occupy as that space renders inside and outside the bounds of the CSS box (usually due to scrolling, 2D, or 3D transformations).

Reviewer: ```suggestion
The `local` keyword enables the position of background images to be fixed relative to the element’s contents. Background images now move along the space they occupy as that space renders inside and outside the bounds of the CSS box (usually due to scrolling, 2D, or 3D transformations).
```

{% Codepen {
  user: 'web-dot-dev',
  id: 'WNExZmK',
  tab: 'result'
} %}

> ## Background Origin

Reviewer: ```suggestion
## Background origin
```

The `background-origin` property enables you to modify the area of backgrounds associated with a particular box. The values the property accepts correspond to the `border-box` , `padding-box`, and `content-box` regions of a box .

Try these options out using the following demo:

{% Codepen {
  user: 'web-dot-dev',
  id: 'ExvyXeZ',
  tab: 'result'
} %}

> ## Background Clip

Reviewer: ```suggestion
## Background clip

```

Like `background-origin` the regions that can be specified are `border-box`, `padding-box`, and `content-box` corresponding to where a CSS background layer can be rendered. When these keywords are used, any rendering of the background further than the region specified will be cropped or clipped.

{% Codepen {
  user: 'web-dot-dev',
  id: 'vYJKZba',
  tab: 'result'
} %}

The `background-clip` property also accepts a `text` keyword that clips the background to be no further than the text within the content box.  For this effect to be evident in the actual text within a CSS box, the text must be partially or completely transparent.

A relatively new property, at the time of this writing, Chrome and most browsers require the `-webkit-` prefix to use this property. For up-to-date information about the browsers that support this option for `background-clip`, [check out this compatibility table](https://caniuse.com/background-clip-text)

{% Codepen {
  user: 'web-dot-dev',
  id: 'qBjjweL',
  tab: 'result,css'
} %}

{% Aside 'gotchas' %}
It should be also noted this property is not compatible with `background-clip: text` being simultaneously set on a CSS box.
{% endAside %}

> ## Multiple Backgrounds

Reviewer: ```suggestion
## Multiple backgrounds

```

The only background defined or the last layer is designated the *final background layer* by the browser. Only this layer is allowed to assign a `background-color`.

{% Aside %}
With the unpredictability of the web in mind, a background-image may fail to load. Setting a background-color on the final layer ensures good contrast for text and so on if important background layers fail to load.
{% endAside %}

Multiple layers can be individually configured using most background-related CSS properties that are comma separated, as demonstrated in the code snippet and live demo below.

```css
background-image: url("https://assets.codepen.io/7518/pngaaa.com-1272986.png"),
    url("https://assets.codepen.io/7518/blob.svg"),
    linear-gradient(hsl(191deg, 40%, 86%), hsla(191deg, 96%, 96%, 0.5));
  background-size: contain, cover, auto;
  background-repeat: no-repeat, no-repeat, no-repeat;
  background-position: 50% 50%, 10% 50%, 0% 0%;
  background-origin: padding-box, border-box, content-box;
```

{% Codepen {
  user: 'web-dot-dev',
  id: 'dyRzQBz',
  tab: 'result,css'
} %}

> ## Background Shorthand

Reviewer: ```suggestion
## The background shorthand
```

If you use a single value, it will be applied to both the horizontal and vertical repeats.

The shorthand also has convenient one-word options to make your intent clearer.

`repeat-x` repeats an image only horizontally; this is equivalent to `repeat no-repeat`.

The following demo demonstrates these capabilities of the `background-repeat` property

{% Codepen {
  user: 'web-dot-dev',
  id: 'KKvMmjb',
  tab: 'result'
} %}

## Background position

You may have noticed when some images on the Web are styled with a `background-repeat: no-repeat` declaration, such images are displayed top left of their container.

The initial position of background images is top left. The `background-position`  property allows you to change this behavior by offsetting the image position.

As with `background-repeat`,  the `background-position` property allows you to position images along the x and y axis independently with two values by default.

When CSS lengths and percentages are used, the first parameter corresponds to the horizontal axis while the second parameter corresponds to the vertical axis.

When keywords are only used the order of the keywords does not matter:

> {% Compare 'better'}
> ```

Reviewer: ```suggestion
{% Compare 'better'}

```css
```

background-position: left 50%
```
{% endCompare%}

{% Compare 'better' %}
background-position: top left;
{% endCompare %}

{% Compare 'better' %}

> ```
> background-position: left top;
> ```

Reviewer: ```suggestion

```css
background-position: left top;
```

```

{% CompareCaption %}
Order does not matter for keywords associated with different axes of position
{% endCompareCaption
{% endCompare %}

{% Compare 'Worse' %}

> ```
>   background-position: 50% left;
> ```

Reviewer: ```suggestion

```css
  background-position: 50% left;
```

```

{% CompareCaption %}
When CSS values are used alongside keywords, it must be preceded by a keyword.
{% endCompareCaption %}
{% endCompare%}

> {% Compare 'Worse' %}
> ```
>   background-position: left right;
> ```

Reviewer: ```suggestion
{% Compare 'Worse' %}
```css
  background-position: left right;
```

```

---
title: Backgrounds
description: >
  In this module learn the ways you can style backgrounds of CSS boxes using CSS
audio:

>   title: The CSS Podcast - 053: Background

Reviewer 2: This needs to go in quotes, since the extra ":" is confusing the YAML parser.

Reviewer: ```suggestion
  title: "The CSS Podcast - 053: Background"
```

  src: https://traffic.libsyn.com/secure/thecsspodcast/TCP052_v1.mp3?dest-id=1891556
  thumbnail: image/foR0vJZKULb5AGJExlazy1xYDgI2/ECDb0qa4TB7yUsHwBic8.png
authors:
 - lozandier

> date: 2021-9-15

Reviewer 2: This probably works, but can you write 2021-09-15 just in case?

Reviewer: ```suggestion
date: 2021-11-09
```

---

# Backgrounds

Behind every CSS box is a specialized layer behind the content called the background layer. CSS provides a variety of ways to make meaningful changes to it–including allowing multiple layers.

This module overviews the ways CSS authors can control the styles of this space.

Background layers are furthest from the user, rendered behind the contents of a CSS box starting from its `padding-box` region. This enables the background layer to not overlap with borders at all.

{% Codepen {
  user: 'argyleink',
  id: 'BaLedvd',
  tab: 'css,result'
} %}

By default, the background layer is directly behind the `border-box` region of a CSS box, a border specified for a CSS box is therefore rendered on top of the background layer, even when a background layer is extended to cover the `border-box`.

## Background Color

> One of the simplest effects you can apply to a background layer is set a [valid CSS color](). The initial value of `background-color` is `transparent`, which allows the contents of a parent to be seen through a CSS box.

Reviewer 2: missing link in ()

> A color set on a background layer sits behind other things painted on a CSS box. This includes borders and outlines (editor’s note: Will deliberately set background-origin to border-box). The demo below demonstrates this.

Reviewer 2: Remove "Editor's note" before publish

{% Codepen {
  user: 'web-dot-dev',
  id: 'XWgRRMQ',
  tab: 'css,result'
} %}

A `background-color` is also guaranteed to sit behind other layers you can create with background-related CSS properties such as a background-image. More on that next.

## Background Image

On top of the `background-color` is a background image, added with the `background-image` property. A `background-image` accepts the following:

* A URL or data URLI using the `url` CSS function.
* An image dynamically created by a gradient CSS function.

### Setting a background-image with the `url` CSS function.

The `url()` CSS  function enables CSS authors to reference a resource as an absolute URL, relative URL, or embed a resource itself in a CSS document by referencing the resource as a  data URI.

The following demo demonstrates this:
{% Codepen {
  user: 'web-dot-dev',
  id: 'JjJNNro',
  tab: 'css,result'
} %}

> {% Aside %}

Reviewer 2: I don't think you close this Aside anywhere. Add `{% endAside %}` when it's done.

```
#someElement {
  background-image: url('https://images.unsplash.com/photo-1536707076730-0615b7f5b49a');
}
```

Whether in this stylesheet or referenced directly in a browser address bar, `https://images.unsplash.com/photo-1536707076730-0615b7f5b49a` is a URL that can be used to consistently & securely fetch a particular image from images.unsplash.com.

*Relative URLs* are URLs that instruct the browser to attempt to fetch a resource relative to the file the URL was embedded in.

```
#anotherElement {
  background-image: url(‘.../some-relative-path/image.png’)
}
```

The `.../some-relative-path/image.png` URL within the stylesheet enables an image to be reliably fetched assuming the location of both the stylesheet & image does not change.

**Data URIs** are a means of describing a resource as bytes of data embedded where it is described; when used by the URL function, you're effectively “inling” the file.

A data URL represents an image as a base64 encoded string. This leads to often a very long string necessary to represent the image accurately:

> ```

Reviewer 2: should be ```css

> #demoBox {

Reviewer 2: Can you actually cut this off rather than including it? Just add some "..."'s after 80-100 characters.

  }
  ```
{% endAside %}

### Setting a background-image with CSS gradient functions

> Several gradient CSS functions exist to allow you to generate a background-image, when passed two or more colors.

Reviewer 2: Add backticks when you use `background-image` with a dash like this. This even works in the heading.

To learn more about the different types of gradients available, see the [CSS module on gradients](https://web.dev/learn/css/gradients/). You can use any of the gradient types as the value for `background-image`.

Regardless of which gradient CSS function is used, the resulting image is [intrinsically sized](https://web.dev/learn/css/box-model/#content-and-sizing) to match the amount of space a background is taking up behind a CSS box.

{% Codepen {
  user: 'web-dot-dev',
  id: 'JjJNNro',
  tab: 'result'
} %}

## Background Repeat

By default, background images are repeated horizontally and vertically to fill the entire space the background layer occupies.

This can be changed by using the `background-repeat` property. The property offers many options to accommodate a variety of use cases.

* `repeat` repeats an image as much as possible within the space available, cropping as necessary.

{% Codepen {
  user: 'web-dot-dev',
  id: 'NWgjgVP',
  tab: 'result'
} %}

{% Codepen {

>   r

Reviewer 2: is this "r" a typo?

  user: 'lozandier',
  id: 'PojmJyE',
  tab: 'result'
} %}

* `round` repeats an image horizontally and vertically to fit as many instances of the image into the space available, without cropping, compressing, or stretching instances of the image.

{% Codepen {
  user: 'web-dot-dev',
  id: 'NWgjgYQ',
  tab: 'result'
} %}

* `space` repeats an image horizontally and vertically to fit as many instances of the image within the space available to the background layer without cropping—spacing out instances of the image as needed. This behavior will have instances of repeating images touching edges of the space a background layer occupies, with white space evenly distributed between all instances:

> ```

Reviewer 2: Not sure why this is in backticks?

```

* `no-repeat` ensures the image is not repeated.

{% Codepen {
  user: 'web-dot-dev',
  id: 'yLXbzeq',
  tab: 'result'
} %}

{% Codepen {
  user: 'lozandier',
  id: 'vYZmZvV',
  tab: 'result'
} %}

The `background-repeat` property allows you to leverage such options for for the x and y axis independently by default. The first parameter sets the horizontal repeat behavior, and the second parameter sets the vertical repeat behavior.

If you use a single value, it will be applied to both the horizontal and vertical repeats.

The shorthand also has the following convenient one-word options to make clearer your intent.

* `repeat-x`  repeats an image only horizontally; this is equivalent to `repeat no-repeat`.

{% Codepen {
  user: 'web-dot-dev',
  id: 'jOwmaWo',
  tab: 'result'
} %}

* `repeat-y`:  repeats an image only vertically; this is equivalent to `no-repeat repeat`.

{% Codepen {
  user: 'web-dot-dev',
  id: 'oNwWomd',
  tab: 'result'
} %}

The following demo shows all of the one-word options.
{% Codepen {
  user: 'web-dot-dev',
  id: 'powPdMb',
  tab: 'result'
} %}

## Background Position

> You may have noticed that when set to `no-repeat`, the image displayed at the top left .

Reviewer 2: remove space between "left" and "."

This is the default position of images on the background-layer. The `background-position` allows you to change this behavior by offsetting the image position.

As with `background-repeat`,  the `background-position` property allows you to position images along the x and y axis independently with two values by default.

When CSS lengths and percentages are used, the first parameter corresponds to the horizontal axis while the second parameter corresponds to the vertical axis.

When keywords are used the order of the keywords does not matter:

```css
#foo {

>   // Works

Reviewer 2: No // comments in CSS, only in Sass or other parsers. So replace with /* */

```css
/* Works */
background-position: bottom 88% right;

/* Works */
background-position: right bottom 88%;

/* Doesn't work; CSS length value must be preceded by the 'top', 'right', 'bottom', or 'left keywords when using 3 or more parameters */
background-position: 88% bottom right;

/* Works */
background-position: bottom 88% right 33%;

/* Works */
background-position: right 33% bottom 88%;

/* Doesn't work; CSS length value must be preceded by the 'top', 'right', 'bottom', or 'left keywords when using 3 or more parameters */
background-position: 88% 33% bottom left;
```

If `background-position: top left 20%` declaration is applied to a CSS background image, the `20%` value represents the 20% X offset from the left of the CSS box, in addition to the image being placed at the very top of the CSS box.

{% Codepen {
  user: 'web-dot-dev',
  id: 'GREmymK',
  tab: 'result'
} %}

 If `background-position: top 20% left` declaration is applied to a CSS background image, the 20% value represents the 20% Y offset from the top of the CSS box, in addition to the image being at the very left of the CSS box.

{% Codepen {
  user: 'web-dot-dev',
  id: 'rNwmpYJ',
  tab: 'result'
} %}

> When four parameters are used, the two keywords are paired with two values corresponding to an offset against the origins of each keyword specified. If `background-position: bottom 20% right 30%` is applied to a background-image, the background-image is positioned 20% from the bottom of the CSS box and and 30% away from the right of the CSS box.

Reviewer 2: backticks on background-image

{% Codepen {
  user: 'web-dot-dev',
  id: 'jOwwxXy',
  tab: 'result'
} %}

> Here are more examples of using the background-position property using a mix of CSS and keyword values:

Reviewer 2: backticks on `background-position`
